curp/src/server/curp_node.rs

```diff
@@ -323,7 +328,7 @@ impl<C: 'static + Command> CurpNode<C> {
         );
 
         // create curp state machine
-        let (voted_for, entries) = storage.recover().await?;
+        let (voted_for, entries, prefix_sum) = storage.recover().await?;
```

Comment:
Should we recover the prefix sum in the `Log::new()` function not in the storage layer? Because prefix sum seems to be an optimization used by `Log` instead of somthing that the storage layer should be awared of.

Reply:
The prefix sum can be viewed as a batch index,  which comes with the log entries. On the performance ground, I rebuild the prefix sum in the recovery process in that it will read the persisted log entries out and serialize them. By doing so we don't need to iterate the log entries again and calculate the serialized size of every single log entry in `Log::new`. But prefix sum is not a good name? Perhaps, batch_index would be better?

Reply:
I still think it's a better idea to calculate the prefix sum in `log.rs`. The performance cost you mentioned is negligible in that the number of recovered log entries will be quite small with gc enabled and the calculation will be performed only once at startup. But the concept of `prefix_sum`, IMO, is still something that the storage layer should not be aware of. The prefix sum does not come with the log entires, but is calculated from the log entries.